packages/eslint-plugin/src/rules/no-base-to-string.ts

```diff
@@ -211,14 +211,37 @@ export default createRule<Options, MessageIds>({
       return Usefulness.Always;
     }
 
-    function hasBaseTypes(type: ts.Type): type is ts.InterfaceType {
-      return (
-        tsutils.isObjectType(type) &&
+    function getBaseTypesForType(type: ts.Type): readonly ts.Type[] {
+      if (!tsutils.isObjectType(type)) {
+        return [];
+      }
+
+      let interfaceType: ts.InterfaceType | undefined;
+
+      if (tsutils.isTypeReference(type)) {
+        const target = type.target;
+        if (
+          tsutils.isObjectFlagSet(
+            target,
+            ts.ObjectFlags.Interface | ts.ObjectFlags.Class,
+          )
+        ) {
+          interfaceType = target;
+        }
+      } else if (
         tsutils.isObjectFlagSet(
           type,
           ts.ObjectFlags.Interface | ts.ObjectFlags.Class,
         )
-      );
+      ) {
+        interfaceType = type as ts.InterfaceType;
+      }
+
+      if (!interfaceType) {
+        return [];
+      }
+
+      return checker.getBaseTypes(interfaceType);
```

Comment:
[Style] The `let` variable confused me a bit - there are a few points of duplication in the code. What do you think about this vague streamlining attempt?

```ts
const interfaceTarget = tsutils.isTypeReference(type)
  ? type.target
  : type;
const interfaceType =
  tsutils.isObjectFlagSet(
    interfaceTarget,
    ts.ObjectFlags.Interface | ts.ObjectFlags.Class,
  ) && (interfaceTarget as ts.InterfaceType);

if (!interfaceType) {
  return [];
}

return checker.getBaseTypes(interfaceType);
```